tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/ValidateAMQPErrorsTest.java

```diff
@@ -0,0 +1,343 @@
+         server.getBrokerConnections().forEach((t) -> Wait.assertTrue(() -> ((AMQPBrokerConnection) t).isConnecting()));
+
+         if (isSender) {
+            server2.start();
+         } else {
+            server.start();
+         }
+
+         server.getBrokerConnections().forEach((t) -> Wait.assertFalse(() -> ((AMQPBrokerConnection) t).isConnecting()));
+
+         createAddressAndQueues(server);
+         createAddressAndQueues(server2);
+
+         Wait.assertTrue(() -> server.locateQueue(getQueueName()) != null);
+         Wait.assertTrue(() -> server2.locateQueue(getQueueName()) != null);
+
+         Thread.sleep(1000);
+
+         ActiveMQServer serverReceivingConnections = isSender ? server2 : server;
+         Wait.assertEquals(1, serverReceivingConnections.getRemotingService()::getConnectionCount);
+         serverReceivingConnections.getRemotingService().getConnections().forEach((t) -> {
+            try {
+               ActiveMQProtonRemotingConnection connection = (ActiveMQProtonRemotingConnection) t;
+               ConnectionImpl protonConnection = (ConnectionImpl) connection.getAmqpConnection().getHandler().getConnection();
+               Wait.waitFor(() -> protonConnection.linkHead(of(ACTIVE), of(ACTIVE)) != null);
+               connection.getAmqpConnection().runNow(() -> {
+                  Link theLink = protonConnection.linkHead(of(ACTIVE), of(ACTIVE));
+                  theLink.close();
+                  connection.flush();
+               });
+            } catch (Exception e) {
+               e.printStackTrace();
+            }
+         });
+
+         ConnectionFactory cf1 = CFUtil.createConnectionFactory("AMQP", "tcp://localhost:" + AMQP_PORT);
+         ConnectionFactory cf2 = CFUtil.createConnectionFactory("AMQP", "tcp://localhost:" + AMQP_PORT_2);
+
+         try (Connection connection = cf1.createConnection()) {
+            Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+            MessageProducer producer = session.createProducer(session.createQueue(getQueueName()));
+            for (int i = 0; i < 10; i++) {
+               producer.send(session.createTextMessage("message " + i));
+            }
+         }
+
+         // messages should still flow after a disconnect on the link
+         // the server should reconnect as if it was a failure
+         try (Connection connection = cf2.createConnection()) {
+            Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+            MessageConsumer consumer = session.createConsumer(session.createQueue(getQueueName()));
+            connection.start();
+            for (int i = 0; i < 10; i++) {
+               Assert.assertEquals("message " + i, ((TextMessage) consumer.receive(5000)).getText());
+            }
+         }
+
+      } finally {
+         AssertionLoggerHandler.stopCapture();
+      }
+   }
+
+   @Test
+   public void testTimeoutOpenOnReceiver() throws Exception {
```

Comment:
I'd suggest testTimeoutOnSenderOpen since its the broker sender handling being tested.

I'd possibly also allow it to succeed on a retry to ensure it recovers. Right now it looks like it would just repeatedly try and fail while the log scrape and shutdown happens. The test could then verify there were 2 connections to the test server, and the first closes due to the error.

Reply:
I'm adding testReconnectAfterSenderOpenTimeout for the test you suggested.


I added this test because errors such as these would never hit maxAttempts since the connection went OK, but something after connection made it to fail.